### PR TITLE
Modification about speculative_sampling_v2

### PR DESCRIPTION
**Pull request by sf-Liu**
Just as described in [the issue](https://github.com/feifeibear/LLMSpeculativeSampling/issues/29#issue-2404413360). I uploaded an alternative method to fix this misalignment.

### CODE REVIEW

File: sampling/speculative_sampling.py [hunk 1]
```diff
@@ -1,185 +1,210 @@
...
+        temperature (float, optional): Defaults to 1.
+        top_k (int, optional): Defaults to 0.
+        top_p (float, optional): Defaults to 0.
+
+    Returns:
+        torch.Tensor: generated tokens (batch, target_seqlen)
+    """
+
+    def batch_padding(x, gamma):
+        PADDING_CODE = 2
```
Comment (feifeibear) [inline sampling/speculative_sampling.py:135]: you can get the padding code from the model?

File: sampling/speculative_sampling.py [hunk 2]
```diff
@@ -1,185 +1,210 @@
...
+            for i in range(q.shape[1]):
+                q[:, i, :] = norm_logits(q[:, i, :],
+                                         temperature, top_k, top_p)
+            # p  = M_p[prefix + x_0, x_0, .., x_(gamma-1)]
+            x_batch, _ = batch_padding(x, gamma)
+
+            p = target_model(x_batch).logits
+            # for i in range(p.shape[1]):
+            #     p[:,i,:] = norm_logits(p[:,i,:],
+            #                     temperature, top_k, top_p)
```
Comment (feifeibear) [inline sampling/speculative_sampling.py:175]: what is the diff between the comments code and the following code?

File: sampling/speculative_sampling.py [hunk 3]
```diff
@@ -1,185 +1,210 @@
...
-            
-            if is_all_accept:
-                t = sample(p[:, -1, :])
-            
-            prefix = torch.cat((prefix, t), dim=1)
-            pbar.update(n - pbar.n)
-
-    return prefix
-
+import torch
```
Comment (feifeibear) [inline sampling/speculative_sampling.py:1]: could you only change your modification code? I hardly see the updates.

File: sampling/speculative_sampling.py [hunk 4]
```diff
@@ -1,185 +1,210 @@
...
+            is_all_accept = True
+            n = prefix_len - 1
+            for i in range(gamma):
+                if random_seed:
+                    torch.manual_seed(random_seed)
+                r = torch.rand(1, device=p.device)
+                j = x[:, prefix_len + i]
+
+                if r < torch.min(torch.tensor([1], device=q.device),
+                                 p[i, -1, j] / q[:, prefix_len + i - 1, j]):
```
Comment (feifeibear) [inline sampling/speculative_sampling.py:192]: I see the difference is here.

File: sampling/speculative_sampling.py (hunk 1 above)
Comment (feifeibear) [reply sampling/speculative_sampling.py:135]: padding_code = tokenizer.pad_token_id
like this way

File: sampling/speculative_sampling.py [hunk 5]
```diff
@@ -1,185 +1,210 @@
...
+
+        n = prefix_len + gamma - 1
+
+        for i in range(gamma):
+            if random_seed:
+                torch.manual_seed(random_seed)
+            r = torch.rand(1, device=device)
+            j = x[:, prefix_len + i]
+
+            if r > (target_model_cache._prob_history[:, prefix_len + i - 1, j]) / (
```
Comment (feifeibear) [inline sampling/speculative_sampling.py:65]: why not also change to `target_model_cache._prob_history.p[i, -1, j]`

File: sampling/speculative_sampling.py (hunk 1 above)
Comment (sf-Liu) [reply sampling/speculative_sampling.py:135]: Revised in next version.

File: sampling/speculative_sampling.py (hunk 4 above)
Comment (sf-Liu) [reply sampling/speculative_sampling.py:192]: Thanks.

File: sampling/speculative_sampling.py (hunk 5 above)
Comment (sf-Liu) [reply sampling/speculative_sampling.py:65]: I made no change on this function. It's as it was.

**Comment by feifeibear** [comment]
https://github.com/feifeibear/LLMSpeculativeSampling/pull/32

I resubmit a PR. 

1. Only updates the diff rather than cover all the file
2. The original implementation is still retained for users to refer to.
3. Fix some errors 

I prefer to attribute the credit to you in the commit history. Could you please resubmit based on my PR? Alternatively, we can also merge my PR?
